pandas/core/internals/array_manager.py

```diff
@@ -665,18 +665,40 @@ def iset(self, loc: Union[int, slice, np.ndarray], value):
         if lib.is_integer(loc):
             # TODO normalize array -> this should in theory not be needed?
             value = extract_array(value, extract_numpy=True)
+
+            # TODO can we avoid needing to unpack this here? That means converting
+            # DataFrame into 1D array when loc is an integer
             if isinstance(value, np.ndarray) and value.ndim == 2:
+                assert value.shape[1] == 1
                 value = value[0, :]
 
             assert isinstance(value, (np.ndarray, ExtensionArray))
-            # value = np.asarray(value)
-            # assert isinstance(value, np.ndarray)
+            assert value.ndim == 1
             assert len(value) == len(self._axes[0])
             self.arrays[loc] = value
             return
 
-        # TODO
-        raise Exception
+        if isinstance(loc, slice):
```

Comment:
can you do an ifelif to make this easier to understand (e.g. include the integer above).